frontends/common/parseInput.cpp

```diff
@@ -65,3 +74,53 @@ const IR::P4Program* parseP4File(CompilerOptions& options) {
     }
     return result;
 }
+
+const IR::P4Program* parseP4String(const std::string& input,
```

Comment:
Yep. I tried merging them at one point, but wasn't happy with the result - they're quite similar, but the differences are in awkward places. I don't mind trying again, though; it might be that a second attempt would go better.